File: 351004/Dashkevich/Security_lab/src/main/java/com/adashkevich/rest/lab/service/EditorService.java
package com.adashkevich.rest.lab.service;

import com.adashkevich.rest.lab.dto.request.EditorRequestTo;
import com.adashkevich.rest.lab.dto.response.EditorResponseTo;
import com.adashkevich.rest.lab.exception.ConflictException;
import com.adashkevich.rest.lab.exception.NotFoundException;
import com.adashkevich.rest.lab.mapper.EditorMapper;
import com.adashkevich.rest.lab.model.Editor;
import com.adashkevich.rest.lab.model.Role;
import com.adashkevich.rest.lab.repository.EditorRepository;
import com.adashkevich.rest.lab.repository.NewsRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class EditorService {
    private final EditorRepository repo;
    private final NewsRepository newsRepo;
    private final EditorMapper mapper;
    private final PasswordEncoder passwordEncoder;

    public EditorService(EditorRepository repo, NewsRepository newsRepo, EditorMapper mapper, PasswordEncoder passwordEncoder) {
        this.repo = repo;
        this.newsRepo = newsRepo;
        this.mapper = mapper;
        this.passwordEncoder = passwordEncoder;
    }

    public EditorResponseTo create(EditorRequestTo dto) {
        return create(dto, false);
    }

    public EditorResponseTo register(EditorRequestTo dto) {
        return create(dto, true);
    }

    private EditorResponseTo create(EditorRequestTo dto, boolean encodePassword) {
        ensureLoginUnique(dto.login, null);
        Editor entity = mapper.toEntity(dto);
        if (entity.getRole() == null) {
            entity.setRole(Role.CUSTOMER);
        }
        if (encodePassword) {
            entity.setPassword(passwordEncoder.encode(dto.password));
        }
        Editor saved = repo.save(entity);
        return mapper.toResponse(saved);
    }

    public List<EditorResponseTo> getAll() {
        return repo.findAll().stream().map(mapper::toResponse).toList();
    }

    public EditorResponseTo getById(Long id) {
        Editor e = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Editor not found", "40410"));
        return mapper.toResponse(e);
    }

    public Optional<Editor> findByLogin(String login) {
        return repo.findAll().stream()
                .filter(editor -> editor.getLogin().equalsIgnoreCase(login))
                .findFirst();
    }

    public EditorResponseTo update(Long id, EditorRequestTo dto) {
        Editor existing = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Editor not found", "40410"));

        ensureLoginUnique(dto.login, id);

        existing.setLogin(dto.login);
        existing.setPassword(dto.password);
        existing.setFirstname(dto.firstname);
        existing.setLastname(dto.lastname);
        existing.setRole(dto.role == null ? Role.CUSTOMER : dto.role);

        Editor updated = repo.update(id, existing);
        return mapper.toResponse(updated);
    }

    public void delete(Long id) {
        Editor editor = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Editor not found", "40410"));

        boolean hasLinkedNews = newsRepo.findAll().stream()
                .anyMatch(news -> id.equals(news.getEditorId()));

        if (hasLinkedNews) {
            throw new ConflictException("Editor is used by news", "40911");
        }

        repo.deleteById(editor.getId());
    }

    private void ensureLoginUnique(String login, Long selfId) {
        boolean exists = repo.findAll().stream()
                .anyMatch(e -> e.getLogin().equalsIgnoreCase(login)
                        && (selfId == null || !e.getId().equals(selfId)));

        if (exists) {
            throw new ConflictException("Editor login must be unique", "40910");
        }
    }
}

File: 351002/Khmylko/rest-api/publisher/src/Service/EditorService.php
<?php
namespace App\Service;

use App\Exception\ConflictException;
use App\Repository\EditorRepository;
use App\Exception\ApiException;
use App\Exception\ValidationException;
use App\Exception\NotFoundException;

class EditorService {
    private EditorRepository $repo;

    public function __construct(EditorRepository $repo) {
        $this->repo = $repo;
    }

    public function getById(int $id): array {
        $editor = $this->repo->findById($id);
        if (!$editor) {
            throw new NotFoundException('Editor', $id);
        }
        // Убираем пароль из ответа
        unset($editor['password']);
        return $editor;
    }

    public function findAll(int $page = 1, int $limit = 10, string $sortBy = 'id', string $order = 'ASC'): array {
        $editors = $this->repo->findAll([], $sortBy, $order, $page, $limit);
        // Убираем пароли
        foreach ($editors as &$editor) {
            unset($editor['password']);
        }
        return $editors;
    }
    public function getAll(): array {
        $editors = $this->repo->findAll();
        foreach ($editors as &$editor) {
            unset($editor['password']);
        }
        return $editors;
    }
    public function createFromRegistrationRequest(array $data): array {
        $normalized = [
            'login' => (string)($data['login'] ?? ''),
            'password' => (string)($data['password'] ?? ''),
            'firstname' => (string)($data['firstName'] ?? $data['firstname'] ?? ''),
            'lastname' => (string)($data['lastName'] ?? $data['lastname'] ?? ''),
            'role' => (string)($data['role'] ?? 'CUSTOMER'),
        ];
        if (!in_array($normalized['role'], ['ADMIN', 'CUSTOMER'], true)) {
            throw new ValidationException('role must be ADMIN or CUSTOMER');
        }
        return $this->create($normalized);
    }

    public function create(array $data): array {
        if (empty($data['login']) || strlen($data['login']) < 2 || strlen($data['login']) > 64) {
            throw new ValidationException("Login must be 2-64 characters");
        }

        // Валидация пароля
        if (empty($data['password']) || strlen($data['password']) < 8 || strlen($data['password']) > 128) {
            throw new ValidationException("Password must be 8-128 characters");
        }

        // Валидация firstname
        if (empty($data['firstname']) || strlen($data['firstname']) < 2 || strlen($data['firstname']) > 64) {
            throw new ValidationException("Firstname must be 2-64 characters");
        }

        // Валидация lastname
        if (empty($data['lastname']) || strlen($data['lastname']) < 2 || strlen($data['lastname']) > 64) {
            throw new ValidationException("Lastname must be 2-64 characters");
        }

        $existing = $this->repo->findByLogin($data['login']);
        if ($existing) {
            throw new ConflictException("Login '{$data['login']}' already exists");
        }

        // Хешируем пароль
        $data['password'] = password_hash($data['password'], PASSWORD_BCRYPT);

        $editor = $this->repo->create($data);
        unset($editor['password']);
        return $editor;
    }

    public function update(int $id, array $data): array {
        $existing = $this->repo->findById($id);
        if (!$existing) {
            throw new NotFoundException('Editor', $id);
        }

        if (isset($data['login'])) {
            $existing['login'] = $data['login'];
        }
        if (isset($data['firstname'])) {
            $existing['firstname'] = $data['firstname'];
        }
        if (isset($data['lastname'])) {
            $existing['lastname'] = $data['lastname'];
        }
        if (isset($data['role'])) {
            $existing['role'] = $data['role'];
        }
        if (isset($data['password'])) {
            $existing['password'] = password_hash($data['password'], PASSWORD_BCRYPT);
        }

        $editor = $this->repo->update($id, $existing);
        unset($editor['password']);
        return $editor;
    }

    public function delete(int $id): void {
        if (!$this->repo->exists($id)) {
            throw new NotFoundException('Editor', $id);
        }
        $this->repo->delete($id);
    }

    private function validate(array $data): void {
        if (empty($data['login']) || strlen($data['login']) < 2 || strlen($data['login']) > 64) {
            throw new ValidationException("Login must be 2-64 characters");
        }
        if (empty($data['password']) || strlen($data['password']) < 8 || strlen($data['password']) > 128) {
            throw new ValidationException("Password must be 8-128 characters");
        }
        if (empty($data['firstname']) || strlen($data['firstname']) < 2 || strlen($data['firstname']) > 64) {
            throw new ValidationException("Firstname must be 2-64 characters");
        }
        if (empty($data['lastname']) || strlen($data['lastname']) < 2 || strlen($data['lastname']) > 64) {
            throw new ValidationException("Lastname must be 2-64 characters");
        }
    }
}